lib/c_backend/transform.ml

```diff
@@ -685,6 +685,103 @@ and prepend_expr env ~prepend_stmts ~append_stmts (expr: expr_result) =
 
   result
 
+and do_literal_folding op left right original_expr =
+  let open Expression in
+
+  let do_integer_folding int_op left_int right_int =
+    let new_int =
+      match int_op with
+      (* TODO: i64 support *)
+      | BinaryOp.Minus -> Int32.(-) left_int right_int
+      | BinaryOp.Plus -> Int32.(+) left_int right_int
+      | BinaryOp.Div -> Int32.(/) left_int right_int
+      | BinaryOp.Mult -> Int32.( * ) left_int right_int
+      | _ -> failwith "unsupported binary op between integer"
+    in
+    Constant(Literal.Integer(new_int))
+  in
+
+  let do_float_folding float_op left_str right_str =
+    let new_float =
+      let left_f = Float.of_string left_str in
+      let right_f = Float.of_string right_str in
+      match float_op with
+      | BinaryOp.Minus -> Float.(-) left_f right_f
+      | BinaryOp.Plus -> Float.(+) left_f right_f
+      | BinaryOp.Div -> Float.(/) left_f right_f
+      | BinaryOp.Mult -> Float.( * ) left_f right_f
+      | _ -> failwith "unsupported binary op between float"
+    in
+    Constant(Literal.Float(Float.to_string new_float, None))
+  in
+
+  let is_same_expr expr_left expr_right = (expr_left.ty_var = expr_right.ty_var) in
+
+  match (left.spec, right.spec) with
+  | (Constant(Integer left_int), Constant(Integer right_int)) -> (
+    match op with
+    | BinaryOp.Plus
+    | BinaryOp.Minus
+    | BinaryOp.Div
+    | BinaryOp.Mult -> (
+      let new_spec = do_integer_folding op left_int right_int in
+      { left with spec = new_spec }
+    )
+    | _ -> original_expr
+  )
+  | (Constant(Float(left_str, _)), Constant(Float(right_str, _))) -> (
```

Comment:
The name `left_str` is a little confuse, is it a `left_foat`?

Reply:
Content of Literal.Float actually with type string, since Float is `string * char option`.  Just to clarify we should use `Float.of_string` below.